src/backend/distributed/catalog/distobject.c

```diff
@@ -0,0 +1,329 @@
+ *
+ * distobject.c
+ *	  Functions to interact with distributed objects by their ObjectAddress
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/heapam.h"
+#include "access/genam.h"
+#include "access/skey.h"
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/dependency.h"
+#include "catalog/namespace.h"
+#include "catalog/objectaddress.h"
+#include "catalog/pg_namespace.h"
+#include "nodes/makefuncs.h"
+#include "nodes/pg_list.h"
+#include "parser/parse_type.h"
+#include "utils/fmgroids.h"
+#include "utils/regproc.h"
+#include "utils/rel.h"
+#include "utils/builtins.h"
+
+#include "distributed/dist_catalog/distobject.h"
+#include "distributed/dist_catalog/pg_dist_object.h"
+#include "distributed/metadata_cache.h"
+
+
+PG_FUNCTION_INFO_V1(citus_prepare_pg_upgrade_pg_dist_object);
+PG_FUNCTION_INFO_V1(citus_finish_pg_upgrade_pg_dist_object);
+
+
+/*
+ * getObjectAddressByIdentifier maps a citus distributed object address to a postgres object
+ * address
+ */
+ObjectAddress *
+getObjectAddressByIdentifier(Oid classId, const char *identifier)
+{
+	ObjectAddress *address = palloc0(sizeof(ObjectAddress));
+
+	switch (classId)
+	{
+		case NamespaceRelationId:
+		{
+			List *names = NIL;
+			const char *namespaceName = NULL;
+			Oid namespaceOid = InvalidOid;
+
+			names = stringToQualifiedNameList(identifier);
+			Assert(list_length(names) == 1);
+
+			namespaceName = strVal(linitial(names));
+			namespaceOid = get_namespace_oid(namespaceName, false);
+
+			ObjectAddressSet(*address, NamespaceRelationId, namespaceOid);
+
+			return address;
+		};
+
+		default:
+		{
+			ereport(ERROR, (errmsg("unrecognized object class: %u", classId)));
+		}
+	}
+}
+
+
+/*
+ * recordObjectDistributed mark an object as a distributed object in citus.
+ */
+void
+recordObjectDistributed(const ObjectAddress *distAddress)
+{
+	Relation pgDistObject = NULL;
+
+	HeapTuple newTuple = NULL;
+	Datum newValues[Natts_pg_dist_object];
+	bool newNulls[Natts_pg_dist_object];
+
+	/* open system catalog and insert new tuple */
+	pgDistObject = heap_open(DistObjectRelationId(), RowExclusiveLock);
+
+	/* form new tuple for pg_dist_partition */
+	memset(newValues, 0, sizeof(newValues));
+	memset(newNulls, false, sizeof(newNulls));
+
+	newValues[Anum_pg_dist_object_classid - 1] = ObjectIdGetDatum(distAddress->classId);
+	newValues[Anum_pg_dist_object_objid - 1] = ObjectIdGetDatum(distAddress->objectId);
+	newNulls[Anum_pg_dist_object_identifier - 1] = true;
+
+	newTuple = heap_form_tuple(RelationGetDescr(pgDistObject), newValues, newNulls);
+
+	/* finally insert tuple, build index entries & register cache invalidation */
+	CatalogTupleInsert(pgDistObject, newTuple);
+
+	CommandCounterIncrement();
+	heap_close(pgDistObject, NoLock);
+}
+
+
+void
+dropObjectDistributed(const ObjectAddress *address)
```

Comment:
rename to unmarkObjectDistributed, verify this is still used by CREATE TYPE later on, or that all is captured by the citus_drop_trigger